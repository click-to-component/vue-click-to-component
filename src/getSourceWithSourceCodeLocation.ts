import { parseFragment } from "parse5";

function getSourceWithSourceCodeLocation({
  source,
  filePath,
  htmlTags,
}: {
  source: string;
  filePath: string;
  htmlTags: readonly string[];
}) {
  const ast = parseFragment(source, {
    sourceCodeLocationInfo: true,
  });

  let allNodes: any[] = [ast];
  let nodeIndex = 0;
  while (allNodes.length > nodeIndex) {
    allNodes = allNodes.concat(
      allNodes[nodeIndex]?.childNodes || [],
      allNodes[nodeIndex]?.content?.childNodes || [],
    );

    nodeIndex++;
  }

  const startOffsetSet = new Set();

  const sortedNodes = allNodes
    .filter((node) => {
      if (!node?.sourceCodeLocation?.startOffset) {
        return false;
      }

      const { startOffset } = node.sourceCodeLocation;

      if (startOffsetSet.has(startOffset)) {
        return false;
      }

      startOffsetSet.add(startOffset);

      if (!htmlTags.includes(node?.nodeName)) {
        return false;
      }

      // Make sure the length of nodeName is equal to the length of the real nodeName in the source
      // For example, <image /> will be parsed as img in the ast
      // Tag name state: https://html.spec.whatwg.org/multipage/parsing.html#tag-name-state
      // eslint-disable-next-line no-control-regex
      if (!/[\u0009\u000A\u000C\u0020\u002F\u003E\u000D]/.test(source[startOffset + node.nodeName.length + 1])) {
        return false;
      }

      return true;
    })
    .sort(
      (a, b) =>
        b.sourceCodeLocation.startOffset - a.sourceCodeLocation.startOffset,
    );

  let result = source;

  sortedNodes.forEach((node) => {
    const { startOffset, startLine, startCol } = node.sourceCodeLocation;
    const sourceCodeLocation = ` data-__source-code-location="${filePath}:${startLine}:${startCol}" `;
    const insertPos = startOffset + node.nodeName.length + 1;
    result =
      result.substring(0, insertPos) +
      sourceCodeLocation +
      result.substring(insertPos);
  });

  return result;
}

export { getSourceWithSourceCodeLocation };
