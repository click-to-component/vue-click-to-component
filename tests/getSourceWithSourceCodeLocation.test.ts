import { describe, expect, test } from "@jest/globals";
import { getSourceWithSourceCodeLocation } from "../src/getSourceWithSourceCodeLocation";
import htmlTags from "html-tags";

function getSourceWithSourceCodeLocationBySource(source: string) {
  return getSourceWithSourceCodeLocation({
    source,
    filePath: "test.vue",
    htmlTags,
  });
}

describe("getSourceWithSourceCodeLocation", () => {
  test("<image /> should not add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><image /></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });

  test("<image\\r> should not add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><image\r></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });

  test("<img /> should add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><img /></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });

  test("<img/> should add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><img/></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });

  test("<img> should add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><img></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });

  test("<img\\r> should add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><img\r></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });

  test("<img\\t> should add source code location", () => {
    const sourceWithSourceCodeLocation =
      getSourceWithSourceCodeLocationBySource("<template><img\t></template>");

    expect(sourceWithSourceCodeLocation).toMatchSnapshot();
  });
});
