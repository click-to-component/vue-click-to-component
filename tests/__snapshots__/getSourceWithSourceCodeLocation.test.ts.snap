// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSourceWithSourceCodeLocation <image /> should not add source code location 1`] = `"<template><image /></template>"`;

exports[`getSourceWithSourceCodeLocation <image\\r> should not add source code location 1`] = `
"<template><image
></template>"
`;

exports[`getSourceWithSourceCodeLocation <img /> should add source code location 1`] = `"<template><img data-__source-code-location="test.vue:1:11"  /></template>"`;

exports[`getSourceWithSourceCodeLocation <img/> should add source code location 1`] = `"<template><img data-__source-code-location="test.vue:1:11" /></template>"`;

exports[`getSourceWithSourceCodeLocation <img> should add source code location 1`] = `"<template><img data-__source-code-location="test.vue:1:11" ></template>"`;

exports[`getSourceWithSourceCodeLocation <img\\r> should add source code location 1`] = `
"<template><img data-__source-code-location="test.vue:1:11" 
></template>"
`;

exports[`getSourceWithSourceCodeLocation <img\\t> should add source code location 1`] = `"<template><img data-__source-code-location="test.vue:1:11" 	></template>"`;
